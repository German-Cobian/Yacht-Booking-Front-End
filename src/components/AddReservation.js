/* eslint-disable react/jsx-props-no-spreading */
import { useState } from 'react';
import { useForm } from 'react-hook-form';
import { useSelector } from 'react-redux';
import { Link, useParams, useNavigate } from 'react-router-dom';
import { getToken } from '../redux/actions/auth';

const ReservationForm = () => {
  const [error, setError] = useState();
  const { id } = useParams();
  const navigate = useNavigate();
  const { currentUser } = useSelector((state) => state.auth);
  const { register, handleSubmit } = useForm();

  const onFormSubmit = async (data) => {
    const response = await fetch('http://localhost:3001/v1/reservations', {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
        Authorization: getToken(),
      },
      body: JSON.stringify({
        user_id: currentUser.id,
        yacht_id: parseInt(id, 10),
        city: data.city,
        start_date: data.start_date,
        end_date: data.end_date,
      }),
    });

    if (response.ok) {
      navigate('/reservations');
    } else {
      setError('We could not add your reservation.');
    }
  };

  return (
    <main className="d-flex flex-column align-items-center">
      <div className="d-flex flex-column col-8 border border-dark my-5">
        <div className="mx-5">
          <h2 className="my-5">BOOK A RESERVATION</h2>
        </div>
        {error && <p className="mx-5">{error}</p>}
        <form className="mx-5" onSubmit={handleSubmit(onFormSubmit)}>
          <div className="form-group my-3">
            <input className="form-control form-control-lg" type="city" placeholder="City" {...register('city', { required: 'City is required' })} />
          </div>
          <div className="form group my-3">
            <input className="form-control form-control-lg" type="start_date" placeholder="Start date" {...register('start_date', { required: 'Start_date is required' })} />
          </div>
          <div className="form-group my-3">
            <input className="form-control form-control-lg" type="end_date" placeholder="End date" {...register('end_date', { required: 'End-date is required' })} />
          </div>
          <input className="btn btn-primary my-5 mx-5" type="submit" value="Add Reservation" />
          <Link className="" to="/reservations">Your Reservations</Link>
        </form>
      </div>
    </main>
  );
};

export default ReservationForm;
