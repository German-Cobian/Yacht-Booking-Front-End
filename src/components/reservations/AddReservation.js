/* eslint-disable react/jsx-props-no-spreading */
import { useState, useEffect } from 'react';
import { useForm } from 'react-hook-form';
import { useSelector } from 'react-redux';
import { Link, useParams, useNavigate } from 'react-router-dom';
import { ToastContainer, toast } from 'react-toastify';
import { getToken } from '../../redux/actions/auth';
import './reservation-form.css';
import 'react-toastify/dist/ReactToastify.css';

const ReservationForm = () => {
  const [yachtName, setYachtName] = useState('');
  const { id } = useParams();
  const navigate = useNavigate();
  const { currentUser } = useSelector((state) => state.auth);
  const { register, handleSubmit } = useForm();

  useEffect(() => {
    (async () => {
      const response = await fetch(`https://wishyacht-api.herokuapp.com/v1/yachts/${id}`, {
        method: 'GET',
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
          Authorization: getToken(),
        },
      });
      if (response.ok) {
        const data = await response.json();
        console.log(data);
        setYachtName(data.attributes.name);
      }
    })();
  }, []);

  const onFormSubmit = async (data) => {
    const response = await fetch('https://wishyacht-api.herokuapp.com/v1/reservations', {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
        Authorization: getToken(),
      },
      body: JSON.stringify({
        reservation: {
          yacht_id: parseInt(id, 10),
          city: data.city,
          start_date: data.start_date,
          days_number: data.days_number.split(' ')[0],
        },
      }),
    });

    if (response.ok) {
      navigate('/reservations');
      toast.success('Reservation added successfully');
    } else {
      toast.error('Dates preceding today cannot be reserved.');
    }
  };

  return (
    <main className="main-2">
      <div className="effect-2" />
      <div className="showcase-2">
        <div className="mx-5">
          <h2 className="my-5 head-form">BOOK A RESERVATION</h2>
        </div>
        <form className="mx-5" onSubmit={handleSubmit(onFormSubmit)}>
          <div className="form-group d-flex gap-3 flex-wrap">
            <label htmlFor="username">
              Username
              <input className="form-control form-control-lg" type="username" value={currentUser.username} disabled />
            </label>
            <label htmlFor="yacht">
              Yacht
              <input className="form-control form-control-lg" type="yacht" value={yachtName} disabled />
            </label>
          </div>
          <div className="d-flex justify-content-between mt-3 flex-wrap">
            <div className="box">
              <input
                type="date"
                name="Select a date"
                {...register('start_date', { required: 'Start_date is required' })}
              />
            </div>
            <div className="box">
              <select {...register('city', { required: 'Start_date is required' })}>
                <option selected="true" disabled="disabled" label="Port of departure?" />
                <option>Rotterdam</option>
                <option>Buenos Aires</option>
                <option>Los Angeles</option>
                <option>Alexandria</option>
                <option>Lisbon</option>
                <option>Shanghai</option>
              </select>
            </div>
            <div className="box">
              <select {...register('days_number', { required: 'days_number is required' })}>
                <option selected="true" disabled="disabled" label="How many days?" />
                <option>1 day</option>
                <option>2 days</option>
                <option>3 days</option>
                <option>4 days</option>
                <option>5 days</option>
                <option>6 days</option>
                <option>7 days</option>
              </select>
            </div>
          </div>
          <div className="action d-flex justify-content-center gap-3 my-3 flex-wrap">
            <input className="reserve" type="submit" value="Add Reservation" />
            <Link className="my-reserve" to="/reservations">
              Your Reservations
            </Link>
          </div>
        </form>
      </div>
      <ToastContainer />
    </main>
  );
};

export default ReservationForm;
