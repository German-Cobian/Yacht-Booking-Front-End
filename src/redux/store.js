// import { legacy_createStore as createStore, applyMiddleware } from 'redux';
// import { composeWithDevTools } from 'redux-devtools-extension';
// import thunk from 'redux-thunk';
// import rootReducer from './reducers';
// import fetchYachts from './actions/yachtActions';

// export default createStore(
//   rootReducer,
//   composeWithDevTools(applyMiddleware(thunk)),
// );

// store.dispatch(fetchYachts());
