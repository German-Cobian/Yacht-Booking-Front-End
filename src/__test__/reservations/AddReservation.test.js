import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import store from '../../redux/storeSetUp';
import AddReservation from '../../components/reservations/AddReservation';

it('renders correctly', () => {
  const tree = renderer.create(
    <Provider store={store}>
      <Router>
        <AddReservation />
      </Router>
    </Provider>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
