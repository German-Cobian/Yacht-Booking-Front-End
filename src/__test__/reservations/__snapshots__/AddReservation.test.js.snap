// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<main
  className="main-2"
>
  <div
    className="effect-2"
  />
  <div
    className="showcase-2"
  >
    <div
      className="mx-5"
    >
      <h2
        className="my-5 head-form"
      >
        BOOK A RESERVATION
      </h2>
    </div>
    <form
      className="mx-5"
      onSubmit={[Function]}
    >
      <div
        className="form-group d-flex gap-3 flex-wrap"
      >
        <label
          htmlFor="username"
        >
          Username
          <input
            className="form-control form-control-lg"
            disabled={true}
            type="username"
          />
        </label>
        <label
          htmlFor="yacht"
        >
          Yacht
          <input
            className="form-control form-control-lg"
            disabled={true}
            type="yacht"
            value=""
          />
        </label>
      </div>
      <div
        className="d-flex justify-content-between mt-3 flex-wrap"
      >
        <div
          className="box"
        >
          <input
            name="start_date"
            onBlur={[Function]}
            onChange={[Function]}
            type="date"
          />
        </div>
        <div
          className="box"
        >
          <select
            name="city"
            onBlur={[Function]}
            onChange={[Function]}
          >
            <option
              disabled="disabled"
              label="Port of departure?"
              selected="true"
            />
            <option>
              Rotterdam
            </option>
            <option>
              Buenos Aires
            </option>
            <option>
              Los Angeles
            </option>
            <option>
              Alexandria
            </option>
            <option>
              Lisbon
            </option>
            <option>
              Shanghai
            </option>
          </select>
        </div>
        <div
          className="box"
        >
          <select
            name="days_number"
            onBlur={[Function]}
            onChange={[Function]}
          >
            <option
              disabled="disabled"
              label="How many days?"
              selected="true"
            />
            <option>
              1 day
            </option>
            <option>
              2 days
            </option>
            <option>
              3 days
            </option>
            <option>
              4 days
            </option>
            <option>
              5 days
            </option>
            <option>
              6 days
            </option>
            <option>
              7 days
            </option>
          </select>
        </div>
      </div>
      <div
        className="action d-flex justify-content-center gap-3 my-3 flex-wrap"
      >
        <input
          className="reserve"
          type="submit"
          value="Add Reservation"
        />
        <a
          className="my-reserve"
          href="/reservations"
          onClick={[Function]}
        >
          Your Reservations
        </a>
      </div>
    </form>
  </div>
</main>
`;
